### PR TITLE
add new ensure_no_dup script for containerd clusters running "kubenet"

### PR DESCRIPTION
**Pull request by eriksywu**
Context:
Historically kubelet has added a set of ebtable rules to dockershim+kubenet nodes when the kubenet bridge is set to use promiscuous mode. This is due to an issue where there'd be duplicated packets sent when podA is talking to podB on the same node via virtual service IP associated with podB (i.e an app does a ns lookup and coredns gets two A record requests).

### CODE REVIEW

File: parts/linux/cloud-init/nodecustomdata.yml [hunk 1]
```diff
@@ -389,6 +389,53 @@ write_files:
...
+
+      cbr0IP=$(ip addr show cbr0 | grep -Eo "inet ([0-9]*\.){3}[0-9]*" | grep -Eo "([0-9]*\.){3}[0-9]*")
+      if [[ -z "${cbr0IP}" ]]; then
+        echo "cbr0 bridge does not have an ipv4 address...checking again in 5s"
+        sleep 5
+        continue
+      fi
+
+      ebtables -t filter -N AKS-DEDUP # add new AKS-DEDUP chain
+      ebtables -t filter -A OUTPUT -j AKS-DEDUP # add new rule to OUTPUT chain jump to AKS-DEDUP
```
Comment (weinong) [inline parts/linux/cloud-init/nodecustomdata.yml:432]: i think it's better to only append to output after AKS-DEDUP chain is configured

File: parts/linux/cloud-init/nodecustomdata.yml [hunk 2]
```diff
@@ -389,6 +389,53 @@ write_files:
...
+      ebtables -t filter -A AKS-DEDUP -p IPv4 -s ${cbr0MAC} -o veth+ --ip-src ${cbr0IP} -j ACCEPT
+      ebtables -t filter -A AKS-DEDUP -p IPv4 -s ${cbr0MAC} -o veth+ --ip-src ${podSubnetAddr} -j DROP
+      exit 0
+    done
+    exit 1
+    #EOF
```
Comment (weinong) [inline parts/linux/cloud-init/nodecustomdata.yml:438]: remove?

**Review by eriksywu — COMMENTED** [review]
confirmed this issue doesn't effect kubenet+calico, as expected. ie:
dig tcpdump of a dns request from client pod:
`01:47:29.965854 IP networkutils.46048 > kube-dns.kube-system.svc.cluster.local.53: 22929+ [1au] A? microsoft.com. (54)`

matching request against coredns logs on the same node...only a single request is received
└──╼ kks logs coredns-b94d8b788-wz5g2 | grep 22929
[INFO] 10.244.0.10:46048 - 22929 "A IN microsoft.com. udp 54 false 4096" NOERROR qr,rd,ra 187 0.002664795s

File: parts/linux/cloud-init/nodecustomdata.yml [hunk 3]
```diff
@@ -389,6 +389,53 @@ write_files:
...
+      fi
+
+      ebtables -t filter -N AKS-DEDUP # add new AKS-DEDUP chain
+      ebtables -t filter -A AKS-DEDUP -p IPv4 -s ${cbr0MAC} -o veth+ --ip-src ${cbr0IP} -j ACCEPT
+      ebtables -t filter -A AKS-DEDUP -p IPv4 -s ${cbr0MAC} -o veth+ --ip-src ${podSubnetAddr} -j DROP
+      ebtables -t filter -A OUTPUT -j AKS-DEDUP # add new rule to OUTPUT chain jump to AKS-DEDUP
+      exit 0
+    done
+    exit 1
+    #EOF
```
Comment (xuto2) [inline parts/linux/cloud-init/nodecustomdata.yml:438]: Let's move this to a static file

**Review by weinong — APPROVED** [review]
LGTM

File: parts/linux/cloud-init/artifacts/ensure_no_dup.sh [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        continue
+    fi
+
+    ebtables -t filter -N AKS-DEDUP # add new AKS-DEDUP chain
+    ebtables -t filter -A AKS-DEDUP -p IPv4 -s ${cbr0MAC} -o veth+ --ip-src ${cbr0IP} -j ACCEPT
+    ebtables -t filter -A AKS-DEDUP -p IPv4 -s ${cbr0MAC} -o veth+ --ip-src ${podSubnetAddr} -j DROP
+    ebtables -t filter -A OUTPUT -j AKS-DEDUP # add new rule to OUTPUT chain jump to AKS-DEDUP
+    exit 0
+done
+exit 1
```
Comment (xuto2) [inline parts/linux/cloud-init/artifacts/ensure_no_dup.sh:41]: error message?

File: parts/linux/cloud-init/artifacts/ensure_no_dup.sh [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        echo "cbr0 bridge does not have an ipv4 address...checking again in 5s"
+        sleep 5
+        continue
+    fi
+
+    ebtables -t filter -N AKS-DEDUP # add new AKS-DEDUP chain
+    ebtables -t filter -A AKS-DEDUP -p IPv4 -s ${cbr0MAC} -o veth+ --ip-src ${cbr0IP} -j ACCEPT
+    ebtables -t filter -A AKS-DEDUP -p IPv4 -s ${cbr0MAC} -o veth+ --ip-src ${podSubnetAddr} -j DROP
+    ebtables -t filter -A OUTPUT -j AKS-DEDUP # add new rule to OUTPUT chain jump to AKS-DEDUP
+    exit 0
```
Comment (xuto2) [inline parts/linux/cloud-init/artifacts/ensure_no_dup.sh:39]: do we want to print out the rules before exiting?

File: parts/linux/cloud-init/artifacts/kubelet.service [hunk 6]
```diff
@@ -18,6 +18,10 @@ ExecStartPre=/bin/mount --make-shared /var/lib/kubelet
 ExecStartPre=-/sbin/ebtables -t nat --list
 ExecStartPre=-/sbin/iptables -t nat --numeric --list
 
+{{- if and NeedsContainerd IsKubenet }}{{- if not HasCalicoNetworkPolicy}}
```
Comment (xuto2) [inline parts/linux/cloud-init/artifacts/kubelet.service:21]: just curious - does calico network provider not using promsc mode or does it already have dedup rule?

**Pull request merged**